pageserver/src/layered_repository.rs

```diff
@@ -357,30 +340,6 @@ impl Repository for LayeredRepository {
     }
 }
 
-fn shutdown_timeline(
-    tenant_id: ZTenantId,
-    timeline_id: ZTimelineId,
-    timeline: &LayeredTimelineEntry,
-) -> Result<(), anyhow::Error> {
-    match timeline {
-        LayeredTimelineEntry::Local(timeline) => {
-            timeline
-                .upload_relishes
```

Comment:
I did some more thinking about this. The possible states of a timeline are:
```
/// A state of the timeline synchronization with the remote storage.
/// Contains `disk_consistent_lsn` of the corresponding remote timeline (latest checkpoint's disk_consistent_lsn).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum TimelineSyncState {
    /// No further downloads from the remote storage are needed.
    /// The timeline state is up-to-date or ahead of the remote storage one,
    /// ready to be used in any pageserver operation.
    Ready(Lsn),
    /// Timeline is scheduled for downloading, but its current local state is not up to date with the remote storage.
    /// The timeline is not ready to be used in any pageserver operations, otherwise it might diverge its local state from the remote version,
    /// making it impossible to sync it further.
    AwaitsDownload(Lsn),
    /// Timeline was not in the pageserver's local working directory, but was found on the remote storage, ready to be downloaded.
    /// Cannot be used in any pageserver operations due to complete absence locally.
    CloudOnly(Lsn),
    /// Timeline was evicted from the pageserver's local working directory due to conflicting remote and local states or too many errors during the synchronization.
    /// Such timelines cannot have their state synchronized further and may not have the data about remote timeline's disk_consistent_lsn, since eviction may happen
    /// due to errors before the remote timeline contents is known.
    Evicted(Option<Lsn>),
}
```

If we're switching to Ready state, then we don't do anything here, before or after this PR. In all the other states, does it even make sense to perform a final checkpoint before changing the state? I don't think it does. In all of those other states, we're not going to use the local data for anything anymore, so what's the point of the final checkpoint? Does that make sense, @SomeoneToIgnore? So I changed this so that we don't perform the final checkpoint.

I also refactored the code in `set_timeline_states`, moving the code to create the LayeredRepository object so that the `Tenant.repo` field can no longer be None. The None was just a transient state between initializing the Tenant and its LayeredRepository, but now that transient state doesn't happen anymore.

Reply:
Yes, I think there's no need to have the checkpoint indeed, but would be nice to set the upload flag to `false` if we don't do that yet.